### PR TITLE
Github stats for indicators

### PR DESCRIPTION
**Pull request by anthotsang**
https://app.asana.com/0/902343675126966/850764393402017

### CODE REVIEW

File: app/helpers/indicators_helper.rb [hunk 1]
```diff
@@ -26,4 +26,8 @@ def get_buy_sell_class(value)
       'neutral'
     end
   end
+
+  def github_url(repo)
+    "https://www.github.com/#{repo}"
```
Comment (hanchang) [inline app/helpers/indicators_helper.rb]: I think their default is to not include the `www` subdomain so might be good to remove it here as well.

File: app/models/coin.rb [hunk 2]
```diff
@@ -33,7 +33,7 @@ class Coin < ApplicationRecord
   has_many :calendar_event_coins
   has_many :calendar_events, through: :calendar_event_coins
 
-  validates :name, uniqueness: true, presence: true
+  validates :name, presence: true
```
Comment (hanchang) [inline app/models/coin.rb]: Hm why did the unique validation get removed?

File: app/models/coin.rb [hunk 3]
```diff
@@ -276,6 +276,24 @@ def sparkline
     end
   end
 
+  def github_stats(force_refresh: false)
+    return nil unless github_repo.present?
```
Comment (hanchang) [inline app/models/coin.rb]: I think you can just do `return` instead of `return nil` to accomplish the same thing.

File: app/models/coin.rb [hunk 4]
```diff
@@ -276,6 +276,24 @@ def sparkline
...
+
+    Rails.cache.fetch("coins/#{id}/github_stats", force: force_refresh) do
+      response = CoinServices::RetrieveGithubStats.call(coin: self)
+      response.try(:result)
+    end
+  end
+
+  def merge_github_stats(stats_to_merge)
+    return nil unless github_repo.present?
```
Comment (hanchang) [inline app/models/coin.rb]: I think you can just do `return` instead of `return nil` to accomplish the same thing.

File: app/models/coin.rb (hunk 2 above)
Comment (anthotsang) [reply app/models/coin.rb]: There were already violations in the database and as far as I can tell, there's no requirement for it?

File: app/services/coin_services/retrieve_github_stats.rb [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        code_frequency: code_frequency,
+        snapshot: snapshot,
+      }
+    end
+
+    private
+
+    def retrieve_commit_activity_data(repo_path)
+      data = @client.commit_activity_stats repo_path
+      return nil unless data.present?
```
Comment (hanchang) [inline app/services/coin_services/retrieve_github_stats.rb]: Still some `return nil` statements but whatevs

File: app/services/coin_services/retrieve_github_stats.rb [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        {
+          timestamp: timestamp,
+          commits: commits,
+        }
+      end
+    end
+
+    def retrieve_code_frequency_data(repo_path)
+      data = @client.code_frequency_stats repo_path
+      return nil unless data.present?
```
Comment (hanchang) [inline app/services/coin_services/retrieve_github_stats.rb]: Still some `return nil` statements but whatevs

File: app/services/coin_services/retrieve_github_stats.rb [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+          timestamp: weekly_data[0],
+          additions: weekly_data[1],
+          deletions: weekly_data[2],
+        }
+      end
+    end
+
+    def retrieve_repository_stats(repo_path)
+      repository = @client.repo repo_path
+      return nil unless repository.present?
```
Comment (hanchang) [inline app/services/coin_services/retrieve_github_stats.rb]: Still some `return nil` statements but whatevs

**Pull request merged**